v2/blockstore/readonly.go

```diff
@@ -52,11 +52,14 @@ func OpenReadOnly(path string, attachIndex bool) (*ReadOnly, error) {
 
 	var idx index.Index
 	if !v2r.Header.HasIndex() {
+
 		idx, err := index.Generate(v2r.CarV1Reader())
 		if err != nil {
 			return nil, err
 		}
 		if attachIndex {
+			// TODO Shouldn't we close the v2r Reader before we open the same file here and write the Index to it.
+			// Can you write to an already opened file and have an open reader pick it up ?
```

Comment:
This is a good question.

We can't just close the mmap reader, because we want to return it. We'd have to open it again.

Ideally we'd only "open" the file once, but I don't think that's possible: x/exp/mmap does not allow writing.

I think the best solution might be to do a regular old os.OpenFile with O_RDWR to check the header and, if the index is missing, calculate it and write it to the file. This shouldn't be slower than the mmap, given that we're reading the entire file anyway. Plus, it's a bit of an edge case; I imagine the fast path would be that the index already exists.

Then, when we're done making sure that the index exists, we close the O_RDWR file and reopen with a read-only mmap, returning the ReadOnly.

Reply:
@aarshkshah1992 thoughts? if it sounds good, I can implement the above.

To be clear, I'm not sure if writing to the file while it's open as an mmap is valid or not. But either way it seems unnecessary and a bit risky.

Reply:
@mvdan Is this done ?

Reply:
I was waiting for your reply :) If you gave the thumbs up yesterday or hours ago, I didn't notice until now - github doesn't send notification emails for emoji reactions.

I'll file an issue about this and do it, though it doesn't seem to be blocking you.

Reply:
Now tracked at https://github.com/ipld/go-car/issues/120.

Reply:
As per #137, more specifically https://github.com/ipld/go-car/pull/137/commits/71bc303b4ce78eea0e44a1537332933b8e3c9e0d `ReadOnly` blockstore API will no longer modify the path given. Instead, a user is encouraged to use `index.Generate` and `index.Attach` to modify an index-less CAR file, may it be v1 or v2. 

Closing #120 as a result.